### PR TITLE
Add Karma System and Update Guild Background Job System

### PR DESCRIPTION
**Pull request by fireflamemm**
- Switched from a pattern of User Side Validation to Access Provider Side Validation on the Database
- Added new tables for User Cards, and Karma Records
- Moved the Guild Background Jobs from being stored in memory to being stored in the Database
- Added a give karma functionality
- Added a check karma functionality
- Improved user experience when the commands error due to permissions
- Added a check to the `event schedule` functionality that verifies that the bot can send messages and `@everyone` in that channel.

### CODE REVIEW

File: HandyHansel/BotService.cs [hunk 1]
```diff
@@ -81,36 +90,75 @@ public async Task StartAsync()
...
             _timeParser = new Parser(_logger, Parser.ParserType.Time);
+
+            RecurringJob.AddOrUpdate<BotService>(bot =>
+                bot.UpdateKarmas(), "0/1 * * * *");
         }
 
         private async Task UpdateDiscordStatus(DiscordClient sender, ReadyEventArgs e)
         {
-            await _discord.UpdateStatusAsync(new DiscordActivity("all the users in disappointment", ActivityType.Watching));
+            await _discord.UpdateStatusAsync(new DiscordActivity("all the users in anticipation", ActivityType.Watching));
```
Comment (fireflamemm) [inline HandyHansel/BotService.cs:113]: Hmm, 🤔, should probably brainstorm the status some.

File: HandyHansel/BotService.cs [hunk 2]
```diff
@@ -81,36 +90,75 @@ public async Task StartAsync()
...
+                e.Handled = true;
+            }
+        }
+
         private  async Task LogExceptions(CommandsNextExtension c, CommandErrorEventArgs e)
         {
-            await e.Context.Channel.SendMessageAsync("An error occurred");
             try
             {
                 DiscordEmbed commandErrorEmbed = new DiscordEmbedBuilder()
```
Comment (fireflamemm) [inline HandyHansel/BotService.cs:159]:
```suggestion
                int stackTraceLength = e.Exception.StackTrace.Length > 1024 ? 1024 : e.Exception.StackTrace.Length;
                DiscordEmbed commandErrorEmbed = new DiscordEmbedBuilder()
```

File: HandyHansel/BotService.cs [hunk 3]
```diff
@@ -81,36 +90,75 @@ public async Task StartAsync()
...
-                    .AddField("Message", e.Exception.Message)
-                    .AddField("StackTrace", e.Exception.StackTrace);
+                    .AddField("Message", e.Exception.Message ?? "")
+                    .AddField("StackTrace", e.Exception.StackTrace.Substring(0, 1024));
```
Comment (fireflamemm) [inline HandyHansel/BotService.cs:161]:
```suggestion
                    .AddField("StackTrace", e.Exception.StackTrace.Substring(0, stackTraceLength));
```

File: HandyHansel/BotService.cs [hunk 4]
```diff
@@ -250,6 +281,32 @@ private async Task<int> PrefixResolver(DiscordMessage msg)
...
+                    userGuildKarmaRecord.CurrentKarma += (ulong)_rng.Next(1, 3);
+                    _userKarmaAddition.Add(userGuildKarmaRecord);
+                }
+            }
+        }
+
+        public async Task UpdateKarmas()
+        {
+            using IBotAccessProvider provider = _accessBuilder.Build();
+            lock(_karmaLock)
```
Comment (fireflamemm) [inline HandyHansel/BotService.cs:304]: Need to verify whether this will lead to deadlocks or not.

File: HandyHansel/BotService.cs (hunk 1 above)
Comment (fireflamemm) [reply HandyHansel/BotService.cs:113]: @Yulzeworth

File: HandyHansel/Commands/KarmaCommands.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                await context.RespondAsync($"{context.Member.Mention}, you can't give yourself Karma. That would be cheating. You have to earn it from others.");
+                return;
+            }
+            using IBotAccessProvider provider = providerBuilder.Build();
+
+            ulong karmaToAdd = 0;
+
+            for (int i = 0; i < 180; i++)
+            {
+                karmaToAdd += (ulong)_rng.Next(1, 4);
```
Comment (cmsteffey) [inline HandyHansel/Commands/KarmaCommands.cs]: I'm confused. If you're not using i, why not just do (ulong)_rng.Next(180, 720)?

**Review by cmsteffey — APPROVED** [review]
Other than the for loop comment, nothing seems wrong. 👍

File: HandyHansel/Commands/KarmaCommands.cs (hunk 5 above)
Comment (fireflamemm) [reply HandyHansel/Commands/KarmaCommands.cs]: 😂, because that would not give the same distribution.

**Pull request merged**